lib/softoken/kbkdf.c

```diff
@@ -0,0 +1,849 @@
+    CK_OBJECT_CLASS classType = CKO_SECRET_KEY;
+    /* Safe to ignore ret? */
+    sftk_forceAttribute(key, CKA_CLASS, &classType, sizeof(classType));
+    return sftk_forceAttribute(key, CKA_VALUE, key_buffer, key_len);
+}
+
+CK_ULONG kbkdf_GetDerivedKeySize(CK_DERIVED_KEY_PTR derived_key) {
+    /* Precondition: kbkdf_ValidateDerived(...) returns CKR_OK for this key,
+     * which implies that keySize is defined. */
+
+    CK_KEY_TYPE keyType = CKK_GENERIC_SECRET;
+    CK_ULONG keySize = 0;
+
+    for (size_t offset = 0; offset < derived_key->ulAttributeCount; offset++) {
+        CK_ATTRIBUTE_PTR template = derived_key->pTemplate + offset;
+
+        /* Find the two attributes we care about. */
+        if (template->type == CKA_KEY_TYPE) {
+            keyType = *(CK_KEY_TYPE *)template->pValue;
+        } else if (template->type == CKA_VALUE_LEN) {
+            keySize = *(CK_ULONG *)template->pValue;
+        }
+    }
+
+    /* Prefer keySize, if we have it. */
+    if (keySize > 0) {
+        return keySize;
+    }
+
+    /* Else, fall back to this mapping. We know kbkdf_ValidateDerived(...)
+     * passed, so this should return non-zero. */
+    return sftk_MapKeySize(keyType);
+}
+
+CK_RV kbkdf_CreateKey(CK_SESSION_HANDLE hSession, CK_DERIVED_KEY_PTR derived_key, SFTKObject **ret_key) {
+    /* Largely duplicated from NSC_Derive(...) */
+    CK_RV ret = CKR_HOST_MEMORY;
+    SFTKObject *key = NULL;
+    SFTKSlot *slot = sftk_SlotFromSessionHandle(hSession);
+    size_t offset = 0;
+
+    PR_ASSERT(ret_key != NULL);
+    PR_ASSERT(derived_key != NULL);
+    PR_ASSERT(derived_key->phKey != NULL);
+
+    /* Create the new key object for this additional derived key. */
+    key = sftk_NewObject(slot);
+    if (key == NULL) {
+        goto done;
+    }
+
+    /* Setup the key from the provided template. */
+    for (offset = 0; offset < derived_key->ulAttributeCount; offset++) {
+        ret = sftk_AddAttributeType(key, sftk_attr_expand(derived_key->pTemplate + offset));
+        if (ret != CKR_OK) {
+            goto done;
+        }
+    }
+
+    *ret_key = key;
+    ret = CKR_OK;
+
+done:
+    return ret;
+}
+
+CK_RV kbkdf_FinalizeKey(CK_SESSION_HANDLE hSession, CK_DERIVED_KEY_PTR derived_key, SFTKObject *key) {
+    /* Largely duplicated from NSC_Derive(...) */
+    CK_RV ret = CKR_HOST_MEMORY;
+    SFTKSession *session = NULL;
+
+    /* TODO: Figure out why this is needed and what it does that (h)Session
+     *       doesn't do? */
+    SFTKSessionObject *sessionForKey = sftk_narrowToSessionObject(key);
+    PR_ASSERT(sessionForKey != NULL);
+    sessionForKey->wasDerived = PR_TRUE;
+
+    session = sftk_SessionFromHandle(hSession);
+    if (session == NULL) {
+        ret = CKR_HOST_MEMORY;
+        goto done;
+    }
+
+    ret = sftk_handleObject(key, session);
+    *(derived_key->phKey) = key->handle;
+    /* TODO: Figure out why we still set phKey when ret != CKR_OK? */
```

Comment:
Probably shouldn't. The caller won't look at it, and it may not have a real meaning.